### ISSUE TITLE
Locking issues are preventing continued operation with Spanner

### ISSUE THREAD

**Issue by jshannon63** [colopl/laravel-spanner#23]
We consistently have issues with locking related to our spanner db access. Seeming like these are related to queued jobs. The issues tend to disappear when we use the sync driver for the queued job processing. Here is an example of one of our exception traces which shows the result but not the likely cause of the issue.

```
Next Illuminate\Database\QueryException: Failed to open lock file. (SQL: select * from `users` where `id` = @p0 limit 1) in /nix/store/krcnq71n6yh724z8dq11lr68lq7gvbwf-composer-laravel-laravel/vendor/colopl/laravel-spanner/src/Colopl/Spanner/Connection.php:441
Stack trace:
#0 /nix/store/krcnq71n6yh724z8dq11lr68lq7gvbwf-composer-laravel-laravel/vendor/laravel/framework/src/Illuminate/Database/Connection.php(629): Colopl\Spanner\Connection->runQueryCallback('select * from `...', Array, Object(Closure))
#1 /nix/store/krcnq71n6yh724z8dq11lr68lq7gvbwf-composer-laravel-laravel/vendor/colopl/laravel-spanner/src/Colopl/Spanner/Connection.php(250): Illuminate\Database\Connection->run('select * from `...', Array, Object(Closure))
#2 /nix/store/krcnq71n6yh724z8dq11lr68lq7gvbwf-composer-laravel-laravel/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2149): Colopl\Spanner\Connection->select('select * from `...', Array, true)
```
Any help in understanding and overcoming this issue would be appreciated.

**Comment by taka-oyama** [comment]
And are you trying to update the same user multiple times?
And Is this query run within a transaction?

If so, this is probably the correct behavior since Spanner will lock the row if you select it in a read-write transaction.

**Comment by jshannon63** [comment]
Where this is happening is within a transaction, but only a single insert statement within the transaction. It only happens when we have background jobs running. The jobs do not write to the users table, only read from it. If we change the jobs to run synchronous then the issue goes away. It seems we only ever have issues with locking when running jobs, and then we have lock issues elsewhere in the code which may be related.

**Comment by taka-oyama** [comment]
I believe reading will still lock the row in a read-write transaction.

QueryException should have an underlying exception.
I need to see that in order to understand it further.

You might want to try looking at the lock statistics to see what's happening on the server side.

https://cloud.google.com/spanner/docs/introspection/lock-statistics

**Comment by jshannon63** [comment]
We have averted our issue temporarily by sync driver for jobs processing. In our many attempts to utilize jobs for background processing, we have failed with this library due to lock conditions for same row jobs. We even wrote a non overlapping job Middleware to avoid this issue, but at the end of the day, it is really no different than the sync driver. If time allows in the future, I may attempt a correct approach to conquer this problem... but for now it will ne synchronous processing.
